src/vm/moar/dispatchers.nqp

```diff
@@ -2574,6 +2599,69 @@ sub raku-multi-plan(
     }
 }
 
+# Helper sub to check whether a Raku Capture contains a "scalar array" (aka $[])
+sub disambiguate-positional-arg-candidates($candidates, $capture) {
+    my int $num-args := nqp::captureposelems($capture);
+    my $capture-scalar-pos-args := nqp::list_i();
+
+    my int $i;
+    while $i < $num-args {
+        if ! nqp::captureposprimspec($capture, $i) {
+            my $value := nqp::captureposarg($capture, $i);
+            nqp::push_i($capture-scalar-pos-args, $i)
+                if nqp::istype_nd($value, Scalar) # could also be nqp::iscont($value)
+                && nqp::istype($value, Array);
+        }
+        ++$i;
+    }
+    my int $total-capture-scalar-pos-args := nqp::elems($capture-scalar-pos-args);
+
+    $i := 0;
+    my %current;
+    while $i < nqp::elems($candidates) {
+        %current := nqp::atpos($candidates, $i);
+        my $signature := nqp::atkey(%current, 'signature');
+        my @types := nqp::atkey(%current, 'types');
+        my @params := nqp::getattr($signature, Signature, '@!params');
+
+        my $z;
+        my $cand-scalar-pos-params := nqp::list_i();
+        while $z < nqp::elems(@params) {
+            my int $flags := nqp::getattr(nqp::atpos(@params, $z), Parameter, '$!flags');
+            nqp::push_i($cand-scalar-pos-params, $z)
+                if ! nqp::bitand_i($flags, nqp::const::SIG_ELEM_INVOCANT)
+                && ! nqp::bitand_i($flags,
+                    (
+                        nqp::const::SIG_ELEM_ARRAY_SIGIL +|
+                        nqp::const::SIG_ELEM_HASH_SIGIL +|
+                        nqp::const::SIG_ELEM_CODE_SIGIL
+                    ))
+                && nqp::istype(
+                    nqp::hllizefor(nqp::atpos(@types, $z), 'Raku'),
+                    nqp::ifnull(
+                        $Positional,
+                        $Positional :=
+                          nqp::gethllsym('Raku', 'MD_Pos')
+                    ));
+            ++$z;
+        }
+
+        my $total-cand-scalar-pos-params := nqp::elems($cand-scalar-pos-params);
+        if $total-capture-scalar-pos-args == $total-cand-scalar-pos-params {
+            my $w;
+            my int $perfect-match := 1;
+            while $w < $total-capture-scalar-pos-args {
+                $perfect-match := $perfect-match
+                    && nqp::iseq_i(nqp::atpos_i($capture-scalar-pos-args, $w), nqp::atpos_i($cand-scalar-pos-params, $w));
+                ++$w;
+            }
+            return [ %current ] if $perfect-match;
+        }
+        ++$i;
+    }
+    return [ %current ];
```

Comment:
There's an ever so slight overhead in using `return`: don't use it in hot code unless it's needed

Reply:
Fixed.